core/lib/dal/src/data_availability_dal.rs

```diff
@@ -0,0 +1,220 @@
+use zksync_db_connection::{
+    connection::Connection,
+    error::DalResult,
+    instrument::{InstrumentExt, Instrumented},
+};
+use zksync_types::{pubdata_da::DataAvailabilityBlob, L1BatchNumber};
+
+use crate::{
+    models::storage_data_availability::{L1BatchDA, StorageDABlob},
+    Core,
+};
+
+#[derive(Debug)]
+pub struct DataAvailabilityDal<'a, 'c> {
+    pub(crate) storage: &'a mut Connection<'c, Core>,
+}
+
+impl DataAvailabilityDal<'_, '_> {
+    /// Inserts the blob_id for the given L1 batch. If the blob_id is already present,
+    /// verifies that it matches the one provided in the function arguments
+    /// (preventing the same L1 batch from being stored twice)
+    pub async fn insert_l1_batch_da(
+        &mut self,
+        number: L1BatchNumber,
+        blob_id: &str,
+        sent_at: chrono::NaiveDateTime,
+    ) -> DalResult<()> {
+        let update_result = sqlx::query!(
+            r#"
+            INSERT INTO
+                data_availability (l1_batch_number, blob_id, sent_at, created_at, updated_at)
+            VALUES
+                ($1, $2, $3, NOW(), NOW())
+            "#,
+            i64::from(number.0),
+            blob_id,
+            sent_at,
+        )
+        .instrument("insert_l1_batch_da")
+        .with_arg("number", &number)
+        .with_arg("blob_id", &blob_id)
+        .report_latency()
+        .execute(self.storage)
+        .await?;
+
+        if update_result.rows_affected() == 0 {
+            tracing::debug!(
+                "L1 batch #{number}: DA blob_id wasn't updated as it's already present"
+            );
+
+            let instrumentation = Instrumented::new("get_matching_batch_da_blob_id")
+                .with_arg("number", &number)
+                .with_arg("blob_id", &blob_id);
+
+            // Batch was already processed. Verify that existing DA blob_id matches
+            let query = sqlx::query!(
+                r#"
+                SELECT
+                    COUNT(*) AS "count!"
+                FROM
+                    data_availability
+                WHERE
+                    l1_batch_number = $1
+                    AND blob_id = $2
+                "#,
+                i64::from(number.0),
+                blob_id,
+            );
+
+            let matched: i64 = instrumentation
+                .clone()
+                .with(query)
+                .report_latency()
+                .fetch_one(self.storage)
+                .await?
+                .count;
+
+            if matched != 1 {
+                let err = instrumentation.constraint_error(anyhow::anyhow!(
+                        "DA blob_id verification failed. DA blob_id {blob_id} for L1 batch #{number} does not match the expected value"
+                    ));
+                return Err(err);
```

Comment:
1 - in some cases of race conditions, not really possible now because we execute everything sequentially
2 - maybe not for all, but for the ones I checked - it is. in case it isn't - the error wouldn't be critical here because the data is already stored, so it's not a big deal if we couldn't store it for the second time
3 - fixed the wording